### ISSUE TITLE
Writing tests

### ISSUE THREAD

**Issue by dricholm** [flutterdata/flutter_data#32]
Hello!

Is there a recommended way for writing tests? If yes, can the documentation be updated with some examples?

I find it hard/impossible to write tests when using flutter_data.

### Adapter mixins

For example if I want to test the Authentication service from the documentation:

```dart
mixin AuthAdapter on Repository<User> {
  Future<String> login(String email, String password) async {
    final response = await withHttpClient(
      (client) => client.post(
        '$baseUrl/token',
        body: _serializeCredentials(user, password),
        headers: headers,
      ),
    );

    final map = json.decode(response.body);
    return map['token'] as String;
  }
}
```

One approach would be to use a separate function to test for `withHttpClient()`'s `OnRequest<R> onRequest` parameter. But for this `http` package needs to be added to dependencies so `Client` can be mocked with `mockito`.

One other way could be to use the `Response` as a parameter for a separate function. But this also mean adding `http` as dependency. Or create a separate class for the project that maybe includes parts of the `Response`.

Would there be some better way?

### With Blocs

If I want to test a Bloc, which has a Repository as constructor parameter. I would normally use `mockito` to mock the repository:

```dart
// This line will result in an error:
// This class (or a class that this class inherits from) is marked as '@immutable', but one or more of its instance fields aren't final: Mock._givenName, Mock._givenHashCode, Mock._defaultResponsedartmust_be_immutable
class MockUserAdapter extends Mock implements LocalAdapter<User> {}

class MockUserRepository extends Mock implements Repository<User> {}

void main() {
  Repository<User> userRepository;

  setUp(() {
    // Need to initialize adapter and also manager
    userRepository = $UserRepository<User>();
  });
```

I can't seem to create a `Mock` of the `Repository` since it also depends on `LocalAdapter`, which depends on `DataManager`. Is there an easy way to mock the `Repository`?

Thank you!

**Comment by frank06** [comment]
@dricholm sorry, I haven't updated the docs yet with testing practices!

Correct, dependencies one is not interested in testing should be faked or mocked.

### Testing objects that depend on Flutter Data

```dart
class Bloc {
  final Repository<Family> repo;
  Bloc(this.repo);
}
```

First mock the repo:

```dart
class MockFamilyRepository extends Mock implements Repository<Family> {}
```

And test:

```dart
void main() async {
  Bloc bloc;

  setUp(() {
    bloc = Bloc(MockFamilyRepository());
  });

  test('mock', () async {
    when(bloc.repo.findAll())
        .thenAnswer((_) => Future(() => [Family(surname: 'Smith')]));
    final families = await bloc.repo.findAll();
    expect(families, predicate((list) => list.first.surname == 'Smith'));
    verify(bloc.repo.findAll());
  });
}
```

### Testing adapters

It's true that testing custom adapters is a bit more convoluted.

Here is an example. Say we have a `login` on the `Person` repo.

```
@DataRepository([PersonLoginAdapter])
class Person with DataSupportMixin<Person> {
 // ...
}

mixin PersonLoginAdapter on Repository<Person> {
  Future<String> login(String email, String password) async {
    final response = await withHttpClient(
      (client) => client.post(
        '$baseUrl/token',
        body: 'some body',
        headers: headers.castToString(),
      ),
    );

    final map = json.decode(response.body);
    return map['token'] as String;
  }
}
```
In our tests, we'll override `withHttpClient` through a new adapter for testing purposes.

In this case we're not interested in testing the HTTP client, so we'll use `MockClient` (provided in the `http` package) to return a fake `token`:

```dart
class PersonRepository = $PersonRepository with TestLoginAdapter;

mixin TestLoginAdapter on PersonLoginAdapter {
  @override
  Future<R> withHttpClient<R>(onRequest) {
    return onRequest(MockClient((req) {
      return Future(() => http.Response('{ "token": "zzz1" }', 200));
    }));
  }
}
```

Now, the actual test:

```dart
final injection = DataServiceLocator();

// USE ANY DI YOU WANT
// as long as its `locator` is of type `T Function<T>()`

void main() async {
  setUpAll(() {
    injection.register(HiveMock());
    final manager = TestDataManager(injection.locator);
    injection.register<DataManager>(manager);
    injection.register<Repository<Person>>(
      PersonRepository(manager, FakeBox<Person>(), remote: false));
  });

  test('custom adapter', () async {
    var repo = injection.locator<Repository<Person>>() as PersonLoginAdapter;
    var token = await repo.login('email@email.com', 'password');
    expect(token, 'zzz1');
  });
}
```

For `TestDataManager` and `FakeBox` see https://github.com/flutterdata/flutter_data/blob/master/test/unit/setup.dart

The data manager is trivial to copy-paste. `FakeBox` I'd argue is something that should be part of the Hive package.

Let me know if this helps.

**Comment by dricholm** [comment]
@frank06 Thank you for the detailed reply!

Oops, I somehow forgot to use the Mock class of the Flutter Data, so testing objects that depend on Flutter Data is really straightforward.

About adapter testing, I have tried your method it really does work. I think you have one minor typo/error I had to change when creating the Repository: The constructor expects a `LocalAdapter`, not the `DataManager`. So get the `DataManager`, create `LocalAdapter(manager)` and then create `Repository(localAdapter)`.

I also tried improving on your method. So I extracted it into a separate file and made it more reusable. I will paste my code here and comment some of it to better understand it.

test_client_setup.dart:
```dart
class MockDataManager extends Mock implements DataManager {}

final _mockResponses = <MockResponse>[];
bool _debug = false;

class MockResponse {
  final Response response;
  // Method 1
  final Request request;
  // Method 2
  final String requestMethod;
  final String requestUrl;
  final Map<String, dynamic> requestBody;

  MockResponse({
    @required this.response,
    this.request,
    this.requestUrl,
    this.requestMethod,
    this.requestBody,
  });
}

// Create the test adapter to override withHttpClient
mixin TestClientAdapter<T extends DataSupportMixin<T>> on Repository<T> {
  @override
  Future<R> withHttpClient<R>(OnRequest<R> onRequest) {
    return onRequest(
      MockClient(
        (request) {
          if (_mockResponses.isEmpty) {
            throw Exception('No mock response registered, got: $request');
          }
          if (_debug) {
            print('Got request: $request');
          }
          final mockResponse = _mockResponses.removeAt(0);

          // Method 1
          if (mockResponse.request != null) {
            expect(
              request.url.toString(),
              equals(mockResponse.request.url.toString()),
            );
            expect(
              request.method,
              equals(mockResponse.request.method),
            );
            expect(
              request.body,
              equals(mockResponse.request.body),
            );
          }

          // Method 2
          if (mockResponse.requestUrl != null) {
            if (_debug) {
              print('Checking URL with ${mockResponse.requestUrl.toString()}');
            }
            expect(
              request.url.toString(),
              equals(mockResponse.requestUrl),
            );
          }

          if (mockResponse.requestMethod != null) {
            if (_debug) {
              print('Checking method with ${mockResponse.requestMethod}');
            }
            expect(
              request.method,
              equals(mockResponse.requestMethod),
            );
          }

          if (mockResponse.requestBody != null) {
            if (_debug) {
              print('Checking body with ${mockResponse.requestBody}');
            }
            expect(
              request.body,
              equals(jsonEncode(mockResponse.requestBody)),
            );
          }

          return Future.value(mockResponse.response);
        },
      ),
    );
  }
}

void setUpTestClient({bool debug = false}) {
  _debug = debug;
  _mockResponses.clear();
}

void verifyNoMoreResponses() {
  if (_mockResponses.isNotEmpty) {
    throw Exception('No match found for: $_mockResponses');
  }
}

void mockClientResponse(MockResponse mockResponse) {
  _mockResponses.add(mockResponse);
}
```

adapter_test.dart
```dart
class TestAuthRepository = AuthRepository with TestClientAdapter;

void main() {
  const baseUrl = 'http://localhost';
  const path = 'users';
  const email = 'user@company.com';
  const password = 'password12';
  AuthRepository authRepository;

  setUp(() {
    setUpTestClient(debug: true);
    authRepository = TestAuthRepository($UserLocalAdapter(MockDataManager()));
  });

  tearDown(() {
    verifyNoMoreResponses();
  });

  test('should return success', () async {
    mockClientResponse(
      MockResponse(
        response: Response('', HttpStatus.created),
        // Method 1
        request: Request('POST', Uri.http('localhost', '/$path'))
          ..body = jsonEncode({'email': email, 'password': password}),
        // Method 2
        requestUrl: '$baseUrl/$path',
        requestMethod: 'POST',
        requestBody: {'email': email, 'password': password},
      ),
    );

    final result = await authRepository.register(
      email: email,
      password: password,
    );

    expect(
      result,
      // Expect null for demonstration purposes only
      equals(null),
    );
  });
}
```

For my case it works now. I am not sure if there could be problems when using it with a more complex Adapter, so maybe you will notice something. The main idea is to just test the client request/response. So I am mocking the whole DataManager class and overriding the withHttpClient in the test mixin. Some points for using it:

1. First setup the test client in the `setUp()` with `setUpTestClient()`, can also use a debug option to check request matching
2. Inside the test call `mockClientResponse()` with the desired `Response` and `Request`. I created a `MockResponse` class for this. I am not sure which method could be better: passing a `Request` object (Method 1) or passing the `Request` parameters separately (Method 2). In the above code I included both for now, but will remove one if I test it more thoroughly or if you have a recommendation. For now I just tried it with a POST request with JSON body.
3. Call `verifyNoMoreResponses()` in the `tearDown()` to check that all responses have been used.

Maybe some other refactors could be added to make it more readable. But for now I want to ask your opinion about this. Do you see anything that could cause a problem with Flutter Data? Also any ideas to improve upon this?

**Comment by frank06** [comment]
Neat!

I don't have a strong opinion about the design yet, but I think it's a great idea to have a `TestClientAdapter`. I'd say method 2 is better, assuming JSON.

I hope one day we can couple this with something like [Mirage](https://miragejs.com). 

You're right about `LocalAdapter`, but it will no longer be a dependency in the next version! Actually, I'm working right now to make `Locator`/`Box` the only dependencies of Repository.

**Comment by dricholm** [comment]
I haven't tried Mirage yet. I usually used [nock](https://github.com/nock/nock) in JavaScript/TypeScript. Just now I tried searching for something similar and I found that someone wrote nock for Dart also: [pub link](https://pub.dev/packages/nock). I haven't tried it yet, will probably try it out soon. If it works okay no `TestClientAdapter` could be needed.

Decreased dependencies sound good, looking forward to the next version!
